boa/integrations/jupyter/browser.py

```diff
@@ -39,26 +37,77 @@
 nest_asyncio.apply()
 
 
-if not colab_eval_js:
-    # colab creates a new iframe for every call, we need to re-inject it every time
-    # for jupyterlab we only need to do it once
-    install_jupyter_javascript_triggers()
+def _install_javascript_triggers():
+    """Run the ethers and titanoboa_jupyterlab Javascript snippets in the browser."""
+    cur_dir = dirname(realpath(__file__))
+    with open(join(cur_dir, "jupyter.js")) as f:
+        jupyter_js = f.read()
+    prefix = os.getenv("JUPYTERHUB_SERVICE_PREFIX", "..")
+    js = jupyter_js.replace("$$JUPYTERHUB_SERVICE_PREFIX", prefix)
+    display(Javascript(js))
+
+
+class BrowserRPC(RPC):
+    """
+    An RPC object that sends requests to the browser via Javascript.
+    """
+
+    def __init__(self):
+        if not colab_eval_js:
+            # colab creates a new iframe for every call, we need to re-inject it every time
+            # for jupyterlab we only need to do it once
+            _install_javascript_triggers()
+
+    @property
+    def identifier(self) -> str:
+        return type(self).__name__  # every instance does the same
+
+    @property
+    def name(self):
+        return self.identifier
+
+    def fetch(
+        self, method: str, params: Any, timeout_message=RPC_TIMEOUT_MESSAGE
+    ) -> Any:
+        return _javascript_call("rpc", method, params, timeout_message=timeout_message)
+
+    def fetch_multi(
+        self, payloads: list[tuple[str, Any]], timeout_message=RPC_TIMEOUT_MESSAGE
+    ) -> list[Any]:
+        return _javascript_call("multiRpc", payloads, timeout_message=timeout_message)
+
+    def wait_for_tx_receipt(self, tx_hash, timeout: float, poll_latency=1):
+        # we do the polling in the browser to avoid too many callbacks
+        # each callback generates currently 10px empty space in the frontend
+        timeout_ms, pool_latency_ms = timeout * 1000, poll_latency * 1000
+        return _javascript_call(
+            "waitForTransactionReceipt",
+            tx_hash,
+            timeout_ms,
+            pool_latency_ms,
+            timeout_message=RPC_TIMEOUT_MESSAGE,
+        )
 
 
 class BrowserSigner:
     """
     A BrowserSigner is a class that can be used to sign transactions in IPython/JupyterLab.
     """
 
-    def __init__(self, address=None):
+    def __init__(self, address=None, rpc=BrowserRPC()):
```

Comment:
`rpc=None`, otherwise rpc is a global -- https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments

Reply:
The browser is global so thats fine?

Reply:
i don't think so. what if set_browser_env() is invoked twice? the state from the previous invocation carries over no?

Reply:
Yes but so does the browser RPC (window object is global on the browser)
There is also no state attached to the RPC at all

Reply:
I changed the code anyway just to avoid confusion later on. But really it's important to understand that there is no difference between BrowserRPC instances

Reply:
> There is also no state attached to the RPC at all

there could be in the future!

> But really it's important to understand that there is no difference between BrowserRPC instances

i mean i believe you ;), but also this could change in the future. it's more robust (to future maintenance changes) to just skip the timebomb and do the standard `default=None` idiom